pkg/volume/rbd/rbd_util.go

```diff
@@ -224,94 +82,72 @@ func (util *RBDUtil) loadRBD(mounter *rbdMounter, mnt string) error {
+			return nbdMapper, nil
+		} else {
+			glog.V(1).Infof("rbd: the rbd-nbd is not availble, fallbacking to krbd")
+			return &RBDKernel{}, nil
+		}
+	}
+	return nil, fmt.Errorf("unsupported backendType %s", b.BackendType)
 }
 
 func (util *RBDUtil) AttachDisk(b rbdMounter) error {
-	var err error
-	var output []byte
-
-	// create mount point
+	// check the mount point first
 	globalPDPath := b.manager.MakeGlobalPDName(*b.rbd)
 	notMnt, err := b.mounter.IsLikelyNotMountPoint(globalPDPath)
 	// in the first time, the path shouldn't exist and IsLikelyNotMountPoint is expected to get NotExist
 	if err != nil && !os.IsNotExist(err) {
 		return fmt.Errorf("rbd: %s failed to check mountpoint", globalPDPath)
 	}
 	if !notMnt {
+		// it looks the specified directory is already a mount point,
+		// so there is no sense in mapping a new rbd device
 		return nil
 	}
 	if err = os.MkdirAll(globalPDPath, 0750); err != nil {
 		return fmt.Errorf("rbd: failed to mkdir %s, error", globalPDPath)
 	}
 
-	devicePath, found := waitForPath(b.Pool, b.Image, 1)
-	if !found {
-		// modprobe
-		_, err = b.plugin.execCommand("modprobe", []string{"rbd"})
-		if err != nil {
-			return fmt.Errorf("rbd: failed to modprobe rbd error:%v", err)
-		}
-
-		// fence off other mappers
-		if err = util.fencing(b); err != nil {
-			return fmt.Errorf("rbd: image %s is locked by other nodes", b.Image)
-		}
-		// rbd lock remove needs ceph and image config
-		// but kubelet doesn't get them from apiserver during teardown
-		// so persit rbd config so upon disk detach, rbd lock can be removed
-		// since rbd json is persisted in the same local directory that is used as rbd mountpoint later,
-		// the json file remains invisible during rbd mount and thus won't be removed accidentally.
-		util.persistRBD(b, globalPDPath)
-
-		// rbd map
-		l := len(b.Mon)
-		// avoid mount storm, pick a host randomly
-		start := rand.Int() % l
-		// iterate all hosts until mount succeeds.
-		for i := start; i < start+l; i++ {
-			mon := b.Mon[i%l]
-			glog.V(1).Infof("rbd: map mon %s", mon)
-			if b.Secret != "" {
-				output, err = b.plugin.execCommand("rbd",
-					[]string{"map", b.Image, "--pool", b.Pool, "--id", b.Id, "-m", mon, "--key=" + b.Secret})
-			} else {
-				output, err = b.plugin.execCommand("rbd",
-					[]string{"map", b.Image, "--pool", b.Pool, "--id", b.Id, "-m", mon, "-k", b.Keyring})
-			}
-			if err == nil {
-				break
-			}
-			glog.V(1).Infof("rbd: map error %v %s", err, string(output))
-		}
-		if err != nil {
-			return fmt.Errorf("rbd: map failed %v %s", err, string(output))
-		}
-		devicePath, found = waitForPath(b.Pool, b.Image, 10)
-		if !found {
-			return errors.New("Could not map image: Timeout after 10s")
-		}
+	// map a block device
+	mapper, err := createDiskMapper(b)
+	if err != nil {
+		return fmt.Errorf("rbd: cannot create diskMapper: %v", err)
 	}
+	devicePath, err := mapper.MapDisk(b)
+	if err != nil {
+		return fmt.Errorf("rbd: cannot map block device")
```

Comment:
Could you return err detail here, when I test this pr, its hard for me to debug the error.

Reply:
May I working on this patch? This patch is very important for ceph supporting persistent volume.

Reply:
I'd say go for it. Please do credit the author and refer to this PR for sure.

Reply:
I have send two emails to this PR's author, but get no reply from him. so sad!